sdk/dynamicconfig/metric/controller/push/period_matcher.go

```diff
@@ -42,28 +42,27 @@ func (matcher *PeriodMatcher) MarkStart(startTime time.Time) {
 	matcher.startTime = startTime
 }
 
-func (matcher *PeriodMatcher) ConsumeSchedules(sched []*pb.MetricConfigResponse_Schedule) {
+func (matcher *PeriodMatcher) ApplySchedules(sched []*pb.MetricConfigResponse_Schedule) time.Duration {
 	matcher.m.Lock()
 	defer matcher.m.Unlock()
 
 	matcher.sched = sched
 	matcher.metrics = make(map[string]*CollectData)
-}
 
-func (matcher *PeriodMatcher) GetMinPeriod() time.Duration {
-	matcher.m.Lock()
-	defer matcher.m.Unlock()
+	return getExportPeriod(matcher.sched)
```

Comment:
You can work out the getExportPeriods before locking.

Also, I think we should add some error handling to getExportPeriod because the schedules come from outside and may be malicious if the system is hacked. For example, what if there is a 0-second schedule? In that case we should return an error and don't update matcher. And we can do so without grabbing the lock.

For now, you can add a TODO, so I am not going to let this block this PR.

Reply:
Sounds good, changing.

Great point; currently, I think if there's a single 0, the code will ignore it. If there are adjacent zero's, the code will panic. I'll go ahead and make a note to explicitly handle schedules with zero's.